R/zchunk_L224.heat.R

```diff
@@ -65,185 +73,341 @@ module_energy_L224.heat_DISABLED <- function(command, ...) {
+      left_join(calibrated_techs %>%
+                  select(sector, fuel, supplysector, subsector, technology, minicam.energy.input) %>%
+                  distinct, by = c("sector", "fuel")) %>%
+      rename(stub.technology = technology) %>%
+      filter(year %in% MODEL_YEARS) %>%
+      filter(region %in% heat_region$region) -> L224.in_EJ_R_heat_F_Yh
+
+    # L224.StubTechCalInput_heat: calibrated input to district heat
+    L224.in_EJ_R_heat_F_Yh %>%
+      select(LEVEL2_DATA_NAMES[["StubTechYr"]], minicam.energy.input, value) %>%
+      mutate(calibrated.value = round(value, energy.DIGITS_CALOUTPUT)) %>%
+      mutate(year.share.weight = year) %>%
+      mutate(subs.share.weight = if_else(calibrated.value == 0, 0, 1)) %>%
+      mutate(share.weight = subs.share.weight) %>%
+      select(-value) -> L224.StubTechCalInput_heat
+
+    # Secondary output of heat, applied to electricity generation technologies
+    # NOTE: This is complicated. Initially tried using historical information for all model periods that fall within historical time
+    # (i.e. not just the model base years). However for regions like the FSU where historical periods often have very low output of heat
+    # from the district heat sector, and most heat as a secondary output from the electricity sector, the secondary output heat can easily
+    # exceed the demands from the end-use sectors, causing model solution failure. For this reason, the convention applied here is to
+    # use the secondary output of heat from the power sector only in the model base years.
+    L124.heatoutratio_R_elec_F_tech_Yh %>%
+      filter(year %in% MODEL_YEARS) %>%
+      left_join(GCAM_region_names, by = "GCAM_region_ID") %>%
+      left_join(calibrated_techs %>%
+                  select(sector, fuel, supplysector, subsector, technology) %>%
+                  distinct, by = c("sector", "fuel", "technology")) %>%
+      mutate(stub.technology = technology) %>%
+      mutate(secondary.output.name = A24.sector$supplysector) -> L224.heatoutratio_R_elec_F_tech_Yh
+
+    # L224.StubTechSecOut_elec: secondary output of district heat from electricity technologies
+    L224.heatoutratio_R_elec_F_tech_Yh %>%
+      select(LEVEL2_DATA_NAMES[["StubTechYr"]], secondary.output.name, value) %>%
+      mutate(secondary.output = round(value, energy.DIGITS_CALOUTPUT)) %>%
+      select(-value) -> L224.StubTechSecOut_elec
+
+    # Calculate cost adjustment, equal to the output of heat multiplied by the heat price (to minimize the distortion of including the secondary output)
+    L224.StubTechSecOut_elec %>%
+      select(LEVEL2_DATA_NAMES[["StubTechYr"]], secondary.output) %>%
+      mutate(minicam.non.energy.input = "heat plant") %>%
+      mutate(input.cost = round(secondary.output*energy.HEAT_PRICE, energy.DIGITS_COST))-> L224.StubTechCost_elec
+
+    # For gas-electric technologies whose efficiencies are below the default assumptions, this cost needs to be reduced
```

Comment:
Could you add a little more to this comment about how this adjustment is being calculated?

Reply:
@kvcalvin I'm adding a description of what calculation is being done, but I'm not sure why they chose to modify it in this way. Do you have any advice?

Reply:
I don't know why it is done either. We can ask Page if @bpbond thinks it is important.

Reply:
I was mostly interested in a description.

Reply:
I think if no one knows why it's being performed, yes, time to tag @pkyle .

Reply:
OK I'll do a comment and some further explanation. Comment:
The secondary output of heat from CHP in the electric sector can cause the price of the technologies to go very low or negative if the technology cost is not modified to reflect the additional costs of CHP systems (as compared with electricity-only systems). Low technology costs can cause unrealistically low electricity prices in the calibration year, distorting behavior in future years. In this method, costs related to heat production and distribution are backed out from exogenous heat prices and data-derived heat:power ratios.

Further commentary: for a GCAM electric sector technology to have non-zero heat output, this indicates that some portion of the electricity in the region (and fuel type) came from the "MAINCHP" flow: facilities whose primary purpose is to produce heat and power for sale to a third party, as opposed to on-site use. In many cases, these "main activity CHP plants" produce 90% heat and 10% electricity; in other words, heat is the primary saleable commodity, and electricity is secondary. It's a bit odd to have these technologies assigned to the electric power sector to begin with, but that's the nature of having mappings that aren't dependent on the magnitude of the flow volumes (which is simpler than the alternative).

There are several possible methods of dealing with this cost/price issue. One is to read in a p-multiplier of zero on the secondary output, and leave all costs as they are. That's the simplest, but it's not an accurate depiction of what's actually happening; in a lot of the cases, heat is the primary output and therefore primary revenue stream. Another alternative is that this method--assigning heat-related costs in order to keep power generation costs consistent with electric-only technologies--could be implemented more precisely, by calculating heat prices in each region and time period from exogenous global fossil energy prices (gas, oil, coal), cost adders, regional T&D losses, and the share of each fuel. Doing this would require a whole lot more data being read in and processed by this code chunk, and given that biomass prices in the calibration years float, the derived heat prices still wouldn't be guaranteed equal to the prices observed in a model run. The method implemented is a middle ground between zeroing out all heat-related revenues, and performing an ad-hoc estimation of the actual heat prices in each region.